src/components/Configurable.js

```diff
@@ -28,12 +29,15 @@ function Configurable(WrappedComponent: React.Component): React.Component {
       };
     }
 
+    // FIXME: Never set state in componentDidMount
     componentDidMount() {
       this.fillWithDefaults(this.props);
     }
 
     componentWillReceiveProps(nextProps: any) {
-      this.fillWithDefaults(nextProps);
+      if (!ObjectUtils.deepEquals(this.props, nextProps)) {
```

Comment:
Still not crazy about this - it's expensive to do a deepEquals comparison every time any state/props change, but I'll leave it be for the time being and come back to it when I can focus on this problem by itself. This is still an improvement as it breaks the loop where this lifecycle would set state unconditionally, infinitely re-triggering itself.